// @sap-ux/adp-tooling/dist/base/prompt.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.promptGeneratorInput = void 0;
const prompts_1 = __importDefault(require("prompts"));
/**
 * Prompt the user for the required properties for an adaptation project.
 *
 * @param defaults optional default values for the prompts
 * @returns a configuration for the adp writer
 */
function promptGeneratorInput(defaults = {}) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        const app = yield (0, prompts_1.default)([
            {
                type: 'select',
                choices: [
                    { title: "CUSTOMER_BASE" /* flexLayer.CUSTOMER_BASE */, value: "CUSTOMER_BASE" /* flexLayer.CUSTOMER_BASE */ },
                    { title: "VENDOR" /* flexLayer.VENDOR */, value: "VENDOR" /* flexLayer.VENDOR */ }
                ],
                name: 'layer',
                message: 'Flex layer:'
            },
            {
                type: 'text',
                name: 'id',
                message: (_prev, values) => {
                    if (values.layer === "CUSTOMER_BASE" /* flexLayer.CUSTOMER_BASE */) {
                        return 'New adaptation id (CUSTOMER_BASE selected, customer prefix will be automatically added to the id):';
                    }
                    else {
                        return 'New adaptation id:';
                    }
                },
                initial: defaults.id,
                format: (input, values) => {
                    if (values.layer === "CUSTOMER_BASE" /* flexLayer.CUSTOMER_BASE */ && !input.startsWith('customer.')) {
                        return `customer.${input}`;
                    }
                    else {
                        return input;
                    }
                },
                validate: (input) => (input === null || input === void 0 ? void 0 : input.length) > 0
            },
            {
                type: 'text',
                name: 'reference',
                message: 'Original application id:',
                initial: defaults.reference,
                validate: (input) => (input === null || input === void 0 ? void 0 : input.length) > 0
            },
            {
                type: 'text',
                name: 'title',
                message: 'Application title:'
            }
        ]);
        const target = yield (0, prompts_1.default)([
            {
                type: 'text',
                name: 'url',
                message: 'Target system url:',
                initial: defaults.url,
                validate: (input) => (input === null || input === void 0 ? void 0 : input.length) > 0
            },
            {
                type: 'text',
                name: 'client',
                message: 'Client (optional):',
                validate: (input) => (input ? input.length < 4 : true)
            }
        ]);
        const deploy = yield (0, prompts_1.default)([
            {
                type: 'text',
                name: 'package',
                message: 'Deployment package:',
                initial: (_a = defaults.package) !== null && _a !== void 0 ? _a : '$TMP',
                validate: (input) => (input === null || input === void 0 ? void 0 : input.length) > 0
            },
            {
                type: 'text',
                name: 'transport',
                message: 'Transport request (optional):',
                initial: defaults.transport
            }
        ]);
        const options = yield (0, prompts_1.default)([
            {
                type: 'confirm',
                name: 'fioriTools',
                message: 'Enable Fiori tools?',
                initial: defaults.ft !== false,
                validate: (input) => (input === null || input === void 0 ? void 0 : input.length) > 0
            }
        ]);
        return { app, target, options, deploy };
    });
}
exports.promptGeneratorInput = promptGeneratorInput;
