// @sap/cds-compiler/lib/transform/effective/misc.js
'use strict';

const {
  forEachDefinition, forEachMemberRecursively, getArtifactDatabaseNameOf, getElementDatabaseNameOf, applyTransformations,
} = require('../../model/csnUtils');
/**
 * Attach @cds.persistence.name to all artifacts and "things".
 * We could also do it more selectively like we do in forRelationalDb, but: Why? Space maybe?
 * @param {CSN.Model} csn
 * @param {CSN.Options} options
 * @param {object} csnUtils
 */
function attachPersistenceName( csn, options, csnUtils ) {
  const { addStringAnnotationTo } = csnUtils;

  forEachDefinition(csn, (artifact, artifactName) => {
    addStringAnnotationTo('@cds.persistence.name', getArtifactDatabaseNameOf(artifactName, options.sqlMapping, csn, options.sqlDialect), artifact);

    forEachMemberRecursively(artifact, (member, memberName) => addStringAnnotationTo('@cds.persistence.name', getElementDatabaseNameOf(memberName, options.sqlMapping, options.sqlDialect), member), [ 'definitions', artifactName ]);
  });
}

/**
 * Delete the given prop from parent.
 *
 * @param {object} parent
 * @param {string|number} prop
 */
function killProp( parent, prop ) {
  delete parent[prop];
}

/**
 * Remove definitions from the CSN:
 * - types
 * - aspects
 *
 * Remove properties from artifacts:
 * - includes
 * - localized
 * @param {CSN.Model} csn
 * @todo Callback-like architecture and merge with persistence name?
 */
function _removeDefinitionsAndProperties( csn ) {
  const killers = {
    $ignore: (a, b, c, path, parentParent) => {
      const tail = path[path.length - 1];
      delete parentParent[tail];
    },
    kind: (artifact, a, b, path) => {
      if (artifact.kind === 'aspect' || artifact.kind === 'type')
        delete csn.definitions[path[1]];

      else if (artifact['@cds.persistence.skip'] === 'if-unused')
        artifact['@cds.persistence.skip'] = false;
    },
    // Still used in flattenStructuredElements - in db/flattening.js
    _flatElementNameWithDots: killProp,
    // Set when setting default string/binary length - used in copyTypeProperties and fixBorkedElementsOfLocalized
    // to not copy the .length property if it was only set via default
    $default: killProp,
    // Set when we turn UUID into String, checked during generateDraftForHana
    $renamed: killProp,
    // Set when we remove .key from temporal things, used in localized.js
    $key: killProp,
    includes: killProp,
    localized: killProp,
    enum: killProp,
    keys: killProp,
    excluding: killProp, // * is resolved, so has no effect anymore
  };

  applyTransformations(csn, killers, [], { skipIgnore: false });
}


module.exports = {
  attachPersistenceName,
  removeDefinitionsAndProperties: _removeDefinitionsAndProperties,
};
